### PR TITLE
Fix Issue #11850 by extending move-model AST visitor to be able to hook BeforeBody in a Block expression

### PR DESCRIPTION
**Pull request by brmataptos**
### Description

Fix Issue #11850, a bug in `flow_insensitive_checkers` struct `SymbolVisitor` which
should process the `binding` of a `Block` expression (which is the RHS
of a `let` expression's variable assignment) in the outer context,
while processing the `body` in the context with the `let` variable(s)
bound.  This was achieved by extending AST's `ExpData::visit_pre_post`
to `visit_positions` which provides hooks not just `Pre` and `Post`
subexpression traversal, but also at useful control points: `PreBody`
for `Block` expressions, and a few other points which will be useful
in other analyses.

### Test Plan
Added the test case which showed the bug.  Other existing tests also
had shown mis-attributed unused variables which have been corrected by
this PR.

### CODE REVIEW

File: third_party/move/move-compiler-v2/src/flow_insensitive_checkers.rs [hunk 1]
```diff
@@ -99,52 +99,63 @@ impl<'env, 'params> SymbolVisitor<'env, 'params> {
         }
     }
 
-    fn entry(&mut self, post: bool, e: &ExpData) -> bool {
+    fn entry(&mut self, position: VisitorPosition, e: &ExpData) -> bool {
         use ExpData::*;
+        use VisitorPosition::*;
         match e {
             Block(_, pat, _, _) => {
```
Comment (wrwg) [inline third_party/move/move-compiler-v2/src/flow_insensitive_checkers.rs:106]: Love it!

**Review by wrwg — APPROVED** [review]

File: third_party/move/move-model/src/ast.rs [hunk 2]
```diff
@@ -915,98 +925,143 @@ impl ExpData {
...
+            if should_continue {
+                Some(())
+            } else {
+                None
+            }
+        });
+    }
+
+    /// Recursively visits expression, calling visitor for key control points of each sub-expression.
+    /// `visitor(Pre, ...)` will be called before descending into each expresison, and
```
Comment (vineethk) [inline third_party/move/move-model/src/ast.rs:944]:
```suggestion
    /// `visitor(Pre, ...)` will be called before descending into each expression, and
```

File: third_party/move/move-model/src/ast.rs [hunk 3]
```diff
@@ -569,6 +569,16 @@ pub enum RewriteResult {
     Unchanged(Exp),
 }
 
+/// Visitor position
+pub enum VisitorPosition {
+    Pre,
```
Comment (vineethk) [inline third_party/move/move-model/src/ast.rs:574]:
```suggestion
    Pre,  // before visiting any children
```

File: third_party/move/move-model/src/ast.rs [hunk 4]
```diff
@@ -569,6 +569,16 @@ pub enum RewriteResult {
...
+    MidMutate,  // after RHS and before LHS of Mutate expression.
+    BeforeBody, // Before body of Block expression.
+    BeforeThen, // Before then clause of IfElse expression.
+    BeforeElse, // Before else clause of IfElse expression.
+    Post,
```
Comment (vineethk) [inline third_party/move/move-model/src/ast.rs:579]:
```suggestion
    Post,  // after visiting all children
```

File: third_party/move/move-compiler-v2/src/flow_insensitive_checkers.rs [hunk 5]
```diff
@@ -99,52 +99,63 @@ impl<'env, 'params> SymbolVisitor<'env, 'params> {
...
-                    // postorder
-                    for (id, var) in pat.vars() {
-                        self.node_symbol_decl_visitor(post, &id, &var, "local variable");
-                    }
-                    self.seen_uses.exit_scope();
-                }
+                match position {
+                    BeforeBody => self.seen_uses.enter_scope(),
+                    Post => {
+                        // postorder
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/src/flow_insensitive_checkers.rs:110]: Consider removing the comment.

**Review by vineethk — APPROVED** [review]
LGTM!

File: third_party/move/move-model/src/ast.rs (hunk 3 above)
Comment (brmataptos) [reply third_party/move/move-model/src/ast.rs:574]: I think "subexpressions" is better than children, but good idea.

File: third_party/move/move-model/src/ast.rs (hunk 4 above)
Comment (brmataptos) [reply third_party/move/move-model/src/ast.rs:579]: same here, subexpressions instead of children.

Comment (vineethk) [reply third_party/move/move-model/src/ast.rs:579]: +1 for subexpressions instead of children

**Pull request merged**